Add frdm_mcxc242 usb support

michal-smola opened the pull request:
Add USB support for frdm_mcxc242 board:
- Adapt driver for USB IP variant
- Enable USB clock
- Add usb configuration in soc devicetree.
- Add usb to board devicetree, update documentation and board yaml
- Configure tests

michal-smola committed:
drivers: usb: Use VREG only if present for NXP Kinetis usb

Some NXP socs with Kinetis USB do not have USB voltage regulator.
It causes build error in USB samples.
Fix it by enabling and disabling the voltage regulator only if
present.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
soc: nxp mcxc: Enable usb clock

USB clock is not enabled for NXP mcxc series. Enable it.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
dts: nxp mcxc: Add usb configuration

USB configuration is missing in Devicetree for NXP MCX C series.
Add the configuration to common and SoC specific dtsi file.
Delete usb node for SoCs without USB support.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
boards: frdm_mcxc242: Add usb support

mcxc242 has usb device capability. Add usb to board documentation and
to the yaml file. Enable and configure it in board dts.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola committed:
samples: usb: Exclude frdm_mcxc242 platform from dfu sample

frdm_mcxc242 does not support MCUboot. Exclude the platform from dfu
sample.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

tmon-nordic commented on drivers/usb/device/usb_dc_kinetis.c:
```diff
@@ -147,8 +147,11 @@ static inline uint8_t get_bdt_idx(uint8_t ep, uint8_t odd)
 
 static int kinetis_usb_init(void)
 {
+#if (defined(FSL_FEATURE_SIM_OPT_HAS_USB_VOLTAGE_REGULATOR) &&                                     \
```
I believe `IS_ENABLED(FSL_FEATURE_SIM_OPT_HAS_USB_VOLTAGE_REGULATOR)` is equivalent to what you achieved here.

michal-smola replied:
Unfortunately, IS_ENABLED doesn't work with values in parenthesis, which is the case for FSL_FEATURE_SIM_OPT_HAS_USB_VOLTAGE_REGULATOR. I can't change it since it is auto-generated code.

jfischer-no reviewed (changes requested):
I do not think it is right approach to pull FSL_FEATURE macros (in a driver that has no dependencies on the MCUX SDK). It looks more like a job for devicetree to describe the hardware and maybe use regulator driver at some point.

michal-smola committed:
dts: usb: Add no-voltage-regulator property to nxp,kinetis-usbd binding

Some NXP SoC's do not have USB voltage regulator present.
Add property to indicate it. Negative logic is used, because
the regulator is present in great majority of SoC's, and thus the new
property can be added only for SoC's without the regulator.

Signed-off-by: Michal Smola <michal.smola@nxp.com>

michal-smola commented:
@jfischer-no , thank you for review. I have added Devicetree property no-voltage-regulator and used it instead of FSL_FEATURE macro. Please review.

michal-smola commented:
@jfischer-no , could you review, please? I have implemented the requested change.

michal-smola commented on drivers/usb/udc/udc_kinetis.c:
```diff
@@ -976,8 +976,10 @@ static int usbfsotg_init(const struct device *dev)
 	const struct usbfsotg_config *config = dev->config;
 	USB_Type *base = config->base;
 
+#if !DT_INST_PROP(0, no_voltage_regulator)
```
@jfischer-no , thank you for the review. I get your point, but SIM module turns on and off the regulator for all instances. It seems to me, that handling it per instance would add unnecessary complexity without a practical effect. SIM_SOPT1_USBREGEN_MASK is not declared for SoCs without the regulator so it has to be compiled conditionally.
Would you reconsider the requested change?

jfischer-no replied:
If so, please use DT_ANY_INST_HAS_PROP_STATUS_OKAY(no_voltage_requlator).

michal-smola commented on samples/subsys/usb/dfu/sample.yaml:
```diff
@@ -18,6 +18,7 @@ common:
     - teensy41
     - b_u585i_iot02a
     - frdm_kl25z
+    - frdm_mcxc242
```
frdm_mcxc242 is excluded, because MCUboot is not supported. There is no plan to support MCUboot because flash is only 64kB.

jfischer-no commented:
@decsny Please take a look.

danieldegrasse commented:
@jfischer-no can you revisit?
